db/attachment_test.go

```diff
@@ -1359,3 +1359,129 @@ func TestMigrateBodyAttachmentsMergeConflicting(t *testing.T) {
+			}
+			var docID = "doc1"
+
+			ctx.addAllowedAttachments(docID, meta, tt.inputBlipProtocol)
+			requireIsAttachmentAllowedTrue(t, ctx, docID, meta, tt.inputBlipProtocol)
+
+			ctx.removeAllowedAttachments(docID, meta, tt.inputBlipProtocol)
+			requireIsAttachmentAllowedFalse(t, ctx, docID, meta, tt.inputBlipProtocol)
+		})
+	}
+
+	// Single document associated with multiple attachments of similar digests.
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := &BlipSyncContext{}
+			meta := []AttachmentStorageMeta{
+				{digest: "digest1", version: tt.inputAttVersion},
+				{digest: "digest1", version: tt.inputAttVersion},
+			}
+			var docID = "doc1"
+
+			ctx.addAllowedAttachments(docID, meta, tt.inputBlipProtocol)
+			key := allowedAttachmentKey(docID, meta[0].digest, tt.inputBlipProtocol)
+			require.True(t, ctx.isAttachmentAllowed(key))
+
+			ctx.removeAllowedAttachments(docID, meta, tt.inputBlipProtocol)
+			require.False(t, ctx.isAttachmentAllowed(key))
+		})
+	}
+
+	// Multiple documents associated with multiple attachments of different digests.
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := &BlipSyncContext{}
+			meta := []AttachmentStorageMeta{
+				{digest: "digest1", version: tt.inputAttVersion},
+				{digest: "digest2", version: tt.inputAttVersion},
+			}
+			var (
+				docID1 = "doc1"
+				docID2 = "doc2"
+			)
+			ctx.addAllowedAttachments(docID1, meta, tt.inputBlipProtocol)
+			requireIsAttachmentAllowedTrue(t, ctx, docID1, meta, tt.inputBlipProtocol)
+			ctx.addAllowedAttachments(docID2, meta, tt.inputBlipProtocol)
+			requireIsAttachmentAllowedTrue(t, ctx, docID2, meta, tt.inputBlipProtocol)
+
+			ctx.removeAllowedAttachments(docID1, meta, tt.inputBlipProtocol)
+			if tt.inputBlipProtocol == BlipCBMobileReplicationV2 && tt.inputAttVersion == AttVersion1 {
+				requireIsAttachmentAllowedTrue(t, ctx, docID1, meta, tt.inputBlipProtocol)
+			} else {
+				requireIsAttachmentAllowedFalse(t, ctx, docID1, meta, tt.inputBlipProtocol)
+			}
+
+			ctx.removeAllowedAttachments(docID2, meta, tt.inputBlipProtocol)
+			requireIsAttachmentAllowedFalse(t, ctx, docID2, meta, tt.inputBlipProtocol)
+		})
+	}
+
+	// Multiple documents associated with multiple attachments of same digests.
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := &BlipSyncContext{}
+			meta := []AttachmentStorageMeta{
+				{digest: "digest1", version: tt.inputAttVersion},
+				{digest: "digest1", version: tt.inputAttVersion},
+			}
+			var (
+				docID1 = "doc1"
+				docID2 = "doc2"
+			)
+			ctx.addAllowedAttachments(docID1, meta, tt.inputBlipProtocol)
+			requireIsAttachmentAllowedTrue(t, ctx, docID1, meta, tt.inputBlipProtocol)
+			ctx.addAllowedAttachments(docID2, meta, tt.inputBlipProtocol)
+			requireIsAttachmentAllowedTrue(t, ctx, docID2, meta, tt.inputBlipProtocol)
+
+			ctx.removeAllowedAttachments(docID1, meta, tt.inputBlipProtocol)
```

Comment:
We should be checking isAllowedAttachment for docID2 here, after removing docID1.

Reply:
Sure, added checking isAllowedAttachment for docID2 here based on replication versions and attachment versions.